### DISCUSSION TITLE
DB integration working in local system not on vercel

### DISCUSSION THREAD

**Discussion by mastindersingh** [vercel/community#5030] [Help]
### Summary

I am attach
[versel.json](https://github.com/vercel/community/files/13520709/versel.json)
ging files.
Tarans-MacBook-Air:stock-db taran$ cat app.py 
from flask import Flask, render_template, request, redirect, url_for, session
import os
from db import read_stock_purchases, write_stock_purchases, get_user_id, verify_user, create_user
import yfinance as yf
import matplotlib
matplotlib.use('Agg')  # Set the backend to Agg
import matplotlib.pyplot as plt
from io import BytesIO
import base64
from decimal import Decimal
import pandas as pd
import logging
from dotenv import load_dotenv
load_dotenv('.env.development.local')

app = Flask(__name__)
app.secret_key = os.environ.get('bApG1HXBfOeC5JhRj_tvKA', 'your-default-secret-key')

@app.route('/user-portfolio', methods=['GET', 'POST'])
def user_portfolio():
    if 'logged_in' not in session:
        return redirect(url_for('login'))

    user_id = session.get('user_id')
    if not user_id:
        print("User ID not found in session. Redirecting to login.")
        return redirect(url_for('login'))

    if request.method == 'POST':
        ticker = request.form.get('ticker')
        buy_date = request.form.get('buy_date')
        buy_price = request.form.get('buy_price')
        quantity = request.form.get('quantity')

        buy_price = Decimal(buy_price)
        quantity = int(quantity)

        new_stock_data = pd.DataFrame({
            'Ticker': [ticker],
            'BuyDate': [buy_date],
            'BuyPrice': [buy_price],
            'Quantity': [quantity]
        })

        write_stock_purchases(user_id, new_stock_data)

    stock_data = fetch_stock_data(read_stock_purchases(user_id))
    return render_template('user_portfolio.html', stock_data=stock_data)

@app.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        email = request.form.get('email')
        password = request.form.get('password')
        if verify_user(email, password):
            session['logged_in'] = True
            session['email'] = email
            user_id = get_user_id(email)  # Get user_id from the database
            if user_id:
                session['user_id'] = user_id  # Set user_id in session
                print(session)
            return redirect(url_for('index'))
        else:
            return render_template('login.html', error="Invalid credentials.")
    return render_template('login.html')

@app.route('/register', methods=['GET', 'POST'])
def register():
    if request.method == 'POST':
        email = request.form.get('email')
        password = request.form.get('password')
        # Add logic to validate the input data
        create_user(email, password)
        return redirect(url_for('login'))
    return render_template('register.html')

@app.route('/logout')
def logout():
    session.clear()
    return redirect(url_for('login'))

@app.route('/', methods=['GET', 'POST'])
def index():
    if 'logged_in' not in session:
        return redirect(url_for('login'))

    user_id = session.get('user_id')
    if not user_id:
        # Handle the case where user_id is not found - perhaps redirect to login
        print("User ID not found in session. Redirecting to login.")
        return redirect(url_for('login'))

    if request.method == 'POST':
        # Handle POST request logic here
        pass

    # Fetch stock data for the logged-in user
    stock_data = fetch_stock_data(read_stock_purchases(user_id))
    return render_template('stocks.html', stock_data=stock_data)

@app.route('/stocks', methods=['POST'])
def add_stock():
    if 'logged_in' not in session:
        return redirect(url_for('login'))

    user_id = session.get('user_id')
    if not user_id:
        print("User ID not found in session. Redirecting to login.")
        return redirect(url_for('login'))

    ticker = request.form.get('ticker')
    buy_date = request.form.get('buy_date')
    buy_price = request.form.get('buy_price')
    quantity = request.form.get('quantity')
    print(f"Received data: {ticker}, {buy_date}, {buy_price}, {quantity}")

    buy_price = Decimal(buy_price)
    quantity = int(quantity)

    new_stock_data = pd.DataFrame({
        'Ticker': [ticker],
        'BuyDate': [buy_date],
        'BuyPrice': [buy_price],
        'Quantity': [quantity]
    })

    write_stock_purchases(user_id, new_stock_data)

    # Fetch the updated stock purchases DataFrame
    updated_stock_purchases_df = read_stock_purchases(user_id)
    stock_data = fetch_stock_data(updated_stock_purchases_df)

    return render_template('stocks.html', stock_data=stock_data)

def fetch_stock_data(stock_purchases):
    stock_data = {}
    for ticker, group in stock_purchases.groupby('Ticker'):
        total_quantity = group['Quantity'].sum()
        weighted_avg_price = Decimal((group['BuyPrice'] * group['Quantity']).sum()) / Decimal(int(total_quantity))
        tickerData = yf.Ticker(ticker)
        data = tickerData.history(period="max")
        current_price = Decimal(data['Close'].iloc[-1])
        percentage_change = ((current_price - weighted_avg_price) / weighted_avg_price) * 100
        performance = "Up" if current_price > weighted_avg_price else "Down"
        #logging.basicConfig(level=logging.DEBUG)
        #logging.debug(f"Stock data for {ticker}: {stock_data[ticker]}")
        plt.figure(figsize=(10, 4))
        plt.plot(data['Close'])
        plt.title(f"{ticker} Stock Price")
        plt.xlabel("Date")
        plt.ylabel("Price")
        graph = get_graph()

        stock_data[ticker] = {
            'data': data,
            'earliest_buy_date': group['BuyDate'].min(),
            'weighted_avg_price': weighted_avg_price,
            'total_quantity': total_quantity,
            'current_price': current_price,
            'performance': performance,
            'percentage_change': percentage_change,
            'graph': graph
        }

    return stock_data

def get_graph():
    buffer = BytesIO()
    plt.savefig(buffer, format='png')
    buffer.seek(0)
    image_png = buffer.getvalue()
    graph = base64.b64encode(image_png)
    graph = graph.decode('utf-8')
    buffer.close()
    return graph
<<<<<<< HEAD

if __name__ == '__main__':
    app.run(debug=True)

=======
>>>>>>> 13f8dc1a151107f1411ceda7e35b345a225163c2Tarans-MacBook-Air:stock-db taran$ cat app.py 
from flask import Flask, render_template, request, redirect, url_for, session
import os
from db import read_stock_purchases, write_stock_purchases, get_user_id, verify_user, create_user
import yfinance as yf
import matplotlib
matplotlib.use('Agg')  # Set the backend to Agg
import matplotlib.pyplot as plt
from io import BytesIO
import base64
from decimal import Decimal
import pandas as pd
import logging
from dotenv import load_dotenv
load_dotenv('.env.development.local')

app = Flask(__name__)
app.secret_key = os.environ.get('bApG1HXBfOeC5JhRj_tvKA', 'your-default-secret-key')

@app.route('/user-portfolio', methods=['GET', 'POST'])
def user_portfolio():
    if 'logged_in' not in session:
        return redirect(url_for('login'))

    user_id = session.get('user_id')
    if not user_id:
        print("User ID not found in session. Redirecting to login.")
        return redirect(url_for('login'))

    if request.method == 'POST':
        ticker = request.form.get('ticker')
        buy_date = request.form.get('buy_date')
        buy_price = request.form.get('buy_price')
        quantity = request.form.get('quantity')

        buy_price = Decimal(buy_price)
        quantity = int(quantity)

        new_stock_data = pd.DataFrame({
            'Ticker': [ticker],
            'BuyDate': [buy_date],
            'BuyPrice': [buy_price],
            'Quantity': [quantity]
        })

        write_stock_purchases(user_id, new_stock_data)

    stock_data = fetch_stock_data(read_stock_purchases(user_id))
    return render_template('user_portfolio.html', stock_data=stock_data)

@app.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        email = request.form.get('email')
        password = request.form.get('password')
        if verify_user(email, password):
            session['logged_in'] = True
            session['email'] = email
            user_id = get_user_id(email)  # Get user_id from the database
            if user_id:
                session['user_id'] = user_id  # Set user_id in session
                print(session)
            return redirect(url_for('index'))
        else:
            return render_template('login.html', error="Invalid credentials.")
    return render_template('login.html')

@app.route('/register', methods=['GET', 'POST'])
def register():
    if request.method == 'POST':
        email = request.form.get('email')
        password = request.form.get('password')
        # Add logic to validate the input data
        create_user(email, password)
        return redirect(url_for('login'))
    return render_template('register.html')

@app.route('/logout')
def logout():
    session.clear()
    return redirect(url_for('login'))

@app.route('/', methods=['GET', 'POST'])
def index():
    if 'logged_in' not in session:
        return redirect(url_for('login'))

    user_id = session.get('user_id')
    if not user_id:
        # Handle the case where user_id is not found - perhaps redirect to login
        print("User ID not found in session. Redirecting to login.")
        return redirect(url_for('login'))

    if request.method == 'POST':
        # Handle POST request logic here
        pass

    # Fetch stock data for the logged-in user
    stock_data = fetch_stock_data(read_stock_purchases(user_id))
    return render_template('stocks.html', stock_data=stock_data)

@app.route('/stocks', methods=['POST'])
def add_stock():
    if 'logged_in' not in session:
        return redirect(url_for('login'))

    user_id = session.get('user_id')
    if not user_id:
        print("User ID not found in session. Redirecting to login.")
        return redirect(url_for('login'))

    ticker = request.form.get('ticker')
    buy_date = request.form.get('buy_date')
    buy_price = request.form.get('buy_price')
    quantity = request.form.get('quantity')
    print(f"Received data: {ticker}, {buy_date}, {buy_price}, {quantity}")

    buy_price = Decimal(buy_price)
    quantity = int(quantity)

    new_stock_data = pd.DataFrame({
        'Ticker': [ticker],
        'BuyDate': [buy_date],
        'BuyPrice': [buy_price],
        'Quantity': [quantity]
    })

    write_stock_purchases(user_id, new_stock_data)

    # Fetch the updated stock purchases DataFrame
    updated_stock_purchases_df = read_stock_purchases(user_id)
    stock_data = fetch_stock_data(updated_stock_purchases_df)

    return render_template('stocks.html', stock_data=stock_data)

def fetch_stock_data(stock_purchases):
    stock_data = {}
    for ticker, group in stock_purchases.groupby('Ticker'):
        total_quantity = group['Quantity'].sum()
        weighted_avg_price = Decimal((group['BuyPrice'] * group['Quantity']).sum()) / Decimal(int(total_quantity))
        tickerData = yf.Ticker(ticker)
        data = tickerData.history(period="max")
        current_price = Decimal(data['Close'].iloc[-1])
        percentage_change = ((current_price - weighted_avg_price) / weighted_avg_price) * 100
        performance = "Up" if current_price > weighted_avg_price else "Down"
        #logging.basicConfig(level=logging.DEBUG)
        #logging.debug(f"Stock data for {ticker}: {stock_data[ticker]}")
        plt.figure(figsize=(10, 4))
        plt.plot(data['Close'])
        plt.title(f"{ticker} Stock Price")
        plt.xlabel("Date")
        plt.ylabel("Price")
        graph = get_graph()

        stock_data[ticker] = {
            'data': data,
            'earliest_buy_date': group['BuyDate'].min(),
            'weighted_avg_price': weighted_avg_price,
            'total_quantity': total_quantity,
            'current_price': current_price,
            'performance': performance,
            'percentage_change': percentage_change,
            'graph': graph
        }

    return stock_data

def get_graph():
    buffer = BytesIO()
    plt.savefig(buffer, format='png')
    buffer.seek(0)
    image_png = buffer.getvalue()
    graph = base64.b64encode(image_png)
    graph = graph.decode('utf-8')
    buffer.close()
    return graph
<<<<<<< HEAD

if __name__ == '__main__':
    app.run(debug=True)

=======
>>>>>>> 13f8dc1a151107f1411ceda7e35b345a225163c2
Tarans-MacBook-Air:stock-db taran$ cat db.py 
import os
import psycopg2
import pandas as pd
from werkzeug.security import generate_password_hash, check_password_hash

def get_db_connection():
    return psycopg2.connect(
        host=os.environ.get("POSTGRES_HOST"),
        database=os.environ.get("POSTGRES_DATABASE"),
        user=os.environ.get("POSTGRES_USER"),
        password=os.environ.get("POSTGRES_PASSWORD"),
        port=os.environ.get("POSTGRES_PORT", "5432")  # Defaulting to 5432 if not set
    )

def read_stock_purchases(user_id):
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("""
        SELECT s.ticker, sp.buy_date, sp.buy_price, sp.quantity 
        FROM stock_purchases sp
        JOIN stocks s ON sp.stock_id = s.id
        WHERE sp.user_id = %s
    """, (user_id,))
    stock_purchases = cursor.fetchall()
    conn.close()
    return pd.DataFrame(stock_purchases, columns=['Ticker', 'BuyDate', 'BuyPrice', 'Quantity'])

def write_stock_purchases(user_id, new_data):
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        for _, row in new_data.iterrows():
            ticker = row['Ticker']
            # Check if the stock ticker exists
            cursor.execute("SELECT id FROM stocks WHERE ticker = %s", (ticker,))
            result = cursor.fetchone()

            if result:
                stock_id = result[0]
            else:
                # Insert new stock ticker
                cursor.execute("INSERT INTO stocks (ticker) VALUES (%s) RETURNING id", (ticker,))
                stock_id = cursor.fetchone()[0]

            # Now insert into stock_purchases with the stock_id
            cursor.execute(
                "INSERT INTO stock_purchases (user_id, stock_id, buy_date, buy_price, quantity) VALUES (%s, %s, %s, %s, %s)",
                (user_id, stock_id, row['BuyDate'], row['BuyPrice'], row['Quantity'])
            )
        conn.commit()
    except psycopg2.Error as e:
        print(f"Database error: {e}")
    finally:
        conn.close()

def get_user_id(email):
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute("SELECT id FROM users WHERE email = %s", (email,))
        result = cursor.fetchone()
        return result[0] if result else None
    except Exception as e:
        print(f"Error in get_user_id: {e}")
        return None
    finally:
        conn.close()

def create_user(email, password):
    conn = get_db_connection()
    cursor = conn.cursor()
    hashed_password = generate_password_hash(password)
    try:
        cursor.execute("INSERT INTO users (email, password) VALUES (%s, %s)", (email, hashed_password))
        conn.commit()
    except psycopg2.Error as e:
        print(f"Database error: {e}")
    finally:
        conn.close()

def verify_user(email, password):
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT password FROM users WHERE email = %s", (email,))
    user_record = cursor.fetchone()
    conn.close()
    if user_record:
        return check_password_hash(user_record[0], password)
    return False

# Function to add a stock to a user's portfolio
def add_stock_to_user(user_id, stock_id):
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute("INSERT INTO user_stocks (user_id, stock_id) VALUES (%s, %s)", (user_id, stock_id))
        conn.commit()
    except psycopg2.Error as e:
        print(f"Database error: {e}")
    finally:
        conn.close()

def get_user_stocks(user_id):
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute("""
            SELECT s.id, s.ticker, COALESCE(SUM(sp.quantity), 0) as total_quantity
            FROM user_stocks us
            JOIN stocks s ON us.stock_id = s.id
            LEFT JOIN stock_purchases sp ON s.id = sp.stock_id AND sp.user_id = us.user_id
            WHERE us.user_id = %s
            GROUP BY s.id
        """, (user_id,))
        stocks = cursor.fetchall()
        return stocks
    except psycopg2.Error as e:
        print(f"Database error: {e}")
    finally:
        conn.close()

def get_master_user_stocks():
    master_email = 'mastinder@yahoo.com'
    master_user_id = get_user_id(master_email)
    return get_user_stocks(master_user_id) if master_user_id else []

def check_user(email, password):
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute("SELECT password FROM users WHERE email = %s", (email,))
        user_record = cursor.fetchone()
        if user_record and check_password_hash(user_record[0], password):
            return True
        else:
            return False
    except psycopg2.Error as e:
        print(f"Database error: {e}")
        return False
    finally:
        conn.close()

### Steps to Reproduce

It is not working on vercel even after giving requirement.txt file
[requirements.txt](https://github.com/vercel/community/files/13520733/requirements.txt)

**Comment by Asaraspo** [comment]
@mastindersingh did you find a solution to this? Seems like I am encountering the same error.

**Comment by kaleblub** [comment]
@mastindersingh I am wondering what db service you are using? I am thinking that Vercel may not be so easy to connect to external databases besides Railway or its own database service... I am not sure yet though.
I am getting a similar error, about psycopg2, even though it is installed and works.
